packages/flutter_tools/lib/src/run_hot.dart

```diff
@@ -988,3 +922,86 @@ class HotRunner extends ResidentRunner {
+      _packageMap = PackageMap(packagesPath).map;
+    } else {
+      _packageMap = const <String, Uri>{};
+    }
+    if (flutterProject != null && flutterProject.pubspecFile.existsSync()) {
+      final PubspecYaml pubspec = PubspecYaml(flutterProject.directory);
+      final Set<String> relevantDependencies = <String>{};
+      for (PubspecDependency dependency in pubspec.allDependencies) {
+        if (dependency.isDevDependency) {
+          continue;
+        }
+        relevantDependencies.add(dependency.name);
+      }
+      // Remove any packages which were tagged as dev dependenices,
+      // But don't remove the app itself!
+      for (String packageName in _packageMap.keys.toList()) {
+        if (!relevantDependencies.contains(packageName) && packageName != flutterProject.manifest.appName) {
+          _packageMap.remove(packageName);
+          continue;
+        }
+      }
+    }
+    // Remove any packages which are derived from the pub cache.
+    for (String packageName in _packageMap.keys.toList()) {
+      final String path = _packageMap[packageName].path;
+      if ((platform.isWindows && path.contains(_pubCachePathWindows))
+          || path.contains(_pubCachePathLinuxAndWindows)) {
+        _packageMap.remove(packageName);
+      }
+    }
+  }
+
+  // Used to avoid watching pubspec directories. This will not change even with pub upgrade,
+  // because that actually switches the directory and requires a corresponding
+  // update to .packages
+  static const String _pubCachePathLinuxAndWindows = '.pub-cache';
+  static const String _pubCachePathWindows = 'Pub/Cache';
+
+  Map<String, Uri> _packageMap;
+  final Map<String, int> _updateTime = <String, int>{};
+
+  List<String> findInvalidated() {
+    final List<String> invalidatedFiles = <String>[];
+    for (String packageName in _packageMap.keys) {
+      final Uri packageUri =_packageMap[packageName];
+      final String packagePath = packageUri.path;
+      // On windows check for leading `/`
+      if (platform.isWindows && packagePath.startsWith(r'/')) {
+        _scanDirectory(packagePath.substring(1), invalidatedFiles);
+      } else {
+        _scanDirectory(packagePath, invalidatedFiles);
+      }
+    }
+    return invalidatedFiles;
+  }
+
+  void _scanDirectory(String path, List<String> invalidatedFiles) {
+    final Directory directory = fs.directory(path);
+    if (!directory.existsSync()) {
+      return;
+    }
+    for (FileSystemEntity entity in directory.listSync(recursive: true)) {
+      if (entity.path.endsWith('.dart')) {
+        final int oldUpdatedAt = _updateTime[entity.path];
+        final int updatedAt = fs.statSync(entity.path).modified.millisecondsSinceEpoch;
+        if (oldUpdatedAt == null || updatedAt > oldUpdatedAt) {
+          // On windows convert to file uri in expected format.
+          if (platform.isWindows) {
+            final String fileUri = 'file:///${entity.path.replaceAll(r'\', r'/')}';
```

Comment:
Can you avoid this Windows specific path-munging by threading Uris through here instead of Strings? For example, the Uri.file constructor says that it will do this replaceAll(): https://api.dartlang.org/dev/2.2.1-dev.1.0/dart-core/Uri/Uri.file.html. Also fs.directory() will accept a Uri, which I assume is implemented somewhere with Directory.fromUri().

Reply:
I originally did it this way because I noticed the Uri-based approach ends up calling Directory_current frequently. I don't actually know how much that slows down the windows update with the other changes though.

I'll update to use the URI based approach for now and follow up with a second PR if I benchmark as being slower

Reply:
Updated

Reply:
If you measure that using Uri is slower than your own string manipulation, we should follow up with a Dart SDK github issue about that.